File: packages/sdk-typescript/src/skill-registry.ts
// SPDX-License-Identifier: Apache-2.0
import { zodToJsonSchema } from 'zod-to-json-schema'
import type { ZodTypeAny } from 'zod'
import { SamvadError, ErrorCode } from './errors.js'
import type { SkillDef, RegisteredSkill, SkillContext, CommunicationMode, TrustTier } from './types.js'

export interface SkillOptions {
  name: string
  description: string
  input: ZodTypeAny
  output: ZodTypeAny
  modes: CommunicationMode[]
  trust: TrustTier
  allowedPeers?: string[]
  handler: (input: unknown, ctx: SkillContext) => Promise<unknown>
}

export class SkillRegistry {
  private skills = new Map<string, RegisteredSkill>()

  register(id: string, opts: SkillOptions): void {
    const def: SkillDef = {
      id,
      name: opts.name,
      description: opts.description,
      inputSchema: zodToJsonSchema(opts.input) as Record<string, unknown>,
      outputSchema: zodToJsonSchema(opts.output) as Record<string, unknown>,
      modes: opts.modes,
      trust: opts.trust,
      allowedPeers: opts.allowedPeers,
    }
    this.skills.set(id, { def, inputZod: opts.input, handler: opts.handler })
  }

  getDefs(): SkillDef[] {
    return Array.from(this.skills.values()).map(s => s.def)
  }

  getSkill(id: string): RegisteredSkill | undefined {
    return this.skills.get(id)
  }

  async dispatch(
    skillId: string,
    payload: Record<string, unknown>,
    ctx: SkillContext,
  ): Promise<unknown> {
    const skill = this.skills.get(skillId)
    if (!skill) throw new SamvadError(ErrorCode.SKILL_NOT_FOUND, `Skill '${skillId}' not found`)

    const result = skill.inputZod.safeParse(payload)
    if (!result.success) {
      const msg = result.error.errors.map(e => `${e.path.join('.')}: ${e.message}`).join(', ')
      throw new SamvadError(ErrorCode.SCHEMA_INVALID, msg)
    }

    return skill.handler(result.data, ctx)
  }
}

File: packages/sdk-python/src/samvad/skill_registry.py
# SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any

from pydantic import BaseModel, ValidationError

from .errors import ErrorCode, SamvadError
from .types import CommunicationMode, SkillContext, SkillDef, TrustTier

Handler = Callable[[BaseModel, SkillContext], Awaitable[BaseModel]]


@dataclass
class RegisteredSkill:
    definition: SkillDef
    input_model: type[BaseModel]
    output_model: type[BaseModel]
    handler: Handler


class SkillRegistry:
    def __init__(self) -> None:
        self._skills: dict[str, RegisteredSkill] = {}

    def register(
        self,
        *,
        name: str,
        description: str,
        input_schema: type[BaseModel],
        output_schema: type[BaseModel],
        modes: list[CommunicationMode],
        trust: TrustTier,
        handler: Handler,
        allowed_peers: list[str] | None = None,
    ) -> None:
        definition = SkillDef(
            id=name,
            name=name,
            description=description,
            inputSchema=input_schema.model_json_schema(),
            outputSchema=output_schema.model_json_schema(),
            modes=list(modes),
            trust=trust,
            allowedPeers=allowed_peers,
        )
        self._skills[name] = RegisteredSkill(
            definition=definition,
            input_model=input_schema,
            output_model=output_schema,
            handler=handler,
        )

    def get(self, name: str) -> RegisteredSkill | None:
        return self._skills.get(name)

    def to_skill_defs(self) -> list[SkillDef]:
        return [s.definition for s in self._skills.values()]

    async def dispatch(
        self, name: str, payload: dict[str, Any], ctx: SkillContext
    ) -> dict[str, Any]:
        skill = self._skills.get(name)
        if skill is None:
            raise SamvadError(ErrorCode.SKILL_NOT_FOUND, f"skill '{name}' not found")
        try:
            parsed = skill.input_model.model_validate(payload, strict=True)
        except ValidationError as e:
            raise SamvadError(ErrorCode.SCHEMA_INVALID, str(e)) from e
        result = await skill.handler(parsed, ctx)
        return result.model_dump(by_alias=True, exclude_none=True)
